cln_plugin/server.go

```diff
@@ -0,0 +1,432 @@
+	s.mtx.Lock()
+	if s.subscription == nil {
+		log.Printf("Got a new HTLC stream subscription request.")
+	} else {
+		s.mtx.Unlock()
+		log.Printf("Got a HTLC stream subscription request, but subscription " +
+			"was already active.")
+		return fmt.Errorf("already subscribed")
+	}
+
+	sb := &subscription{
+		stream: stream,
+		done:   make(chan struct{}),
+		err:    make(chan error, 1),
+	}
+	s.subscription = sb
+
+	// Notify listeners that a new subscriber is active. Replace the chan with
+	// a new one immediately in case this subscriber is dropped later.
+	close(s.newSubscriber)
+	s.newSubscriber = make(chan struct{})
+	s.mtx.Unlock()
+
+	defer func() {
+		// When the HtlcStream function returns, that means the subscriber will
+		// be gone. Cleanup the subscription so we'll be ready to accept a new
+		// one later.
+		s.removeSubscriptionIfUnchanged(sb, nil)
+	}()
+
+	go func() {
+		// If the context is done, there will be no more connection with the
+		// client. Listen for context done and clean up the subscriber.
+		// Cleaning up the subscriber will make the HtlcStream function exit.
+		// (sb.done or sb.err)
+		<-stream.Context().Done()
+		log.Printf("HtlcStream context is done. Removing subscriber: %v", stream.Context().Err())
+		s.removeSubscriptionIfUnchanged(sb, stream.Context().Err())
+	}()
+
+	select {
+	case <-s.done:
+		log.Printf("HTLC server signalled done. Return EOF.")
+		return io.EOF
+	case <-sb.done:
+		log.Printf("HTLC stream signalled done. Return EOF.")
+		return io.EOF
+	case err := <-sb.err:
+		log.Printf("HTLC stream signalled error. Return %v", err)
+		return err
+	}
+}
+
+// Enqueues a htlc_accepted message for send to the grpc client.
+func (s *server) Send(id string, h *HtlcAccepted) {
+	s.sendQueue <- &htlcAcceptedMsg{
+		id:      id,
+		htlc:    h,
+		timeout: time.Now().Add(s.subscriberTimeout),
+	}
+}
+
+// Receives the next htlc resolution message from the grpc client. Returns id
+// and message. Blocks until a message is available. Returns a nil message if
+// the server is done. This function effectively waits until a subscriber is
+// active and has sent a message.
+func (s *server) Receive() (string, interface{}) {
+	select {
+	case <-s.done:
+		return "", nil
+	case msg := <-s.recvQueue:
+		return msg.id, msg.result
+	}
+}
+
+// Listens to sendQueue for htlc_accepted requests from cln. The message will be
+// held until a subscriber is active, or the subscriber timeout expires. The
+// messages are sent to the grpc client in fifo order.
+func (s *server) listenHtlcRequests() {
+	for {
+		select {
+		case <-s.done:
+			log.Printf("listenHtlcRequests received done. Stop listening.")
+			return
+		case msg := <-s.sendQueue:
+			s.handleHtlcAccepted(msg)
+		}
+	}
+}
+
+// Attempts to send a htlc_accepted message to the grpc client. The message will
+// be held until a subscriber is active, or the subscriber timeout expires.
+func (s *server) handleHtlcAccepted(msg *htlcAcceptedMsg) {
+	for {
```

Comment:
So the for loop is for retry sending the htlc in case of error in the send (stream error), is that correct?
In that case maybe it is better to wrap only the actual code that sends in this loop (sb.stream.Send) ?
Also maybe we want to limit the retries at some point and just use default resolution?
I am concerns of cases where there some infrastructure problem that prevents communication with the subscriber which will results in holding all htlcs.

Reply:
It basically holds the htlc until there is a new subscriber available, or the subscribertimeout expires.
It doesn't really retry. 
- If there is no subscriber, it will wait until there is one
- if there is a subscriber, but it drops, so the send errors, it will wait until there is a new subscriber

You wouldn't expect this to replay very often. Just one retry if the subscriber just dropped basically.

In any case, there is a subscriber timeout. Which means if there is no subscriber after holding the htlc x time, it will return the default resolution. That does mean that all htlc's will be held for that duration if there will be no subscriber at all, yes.

Reply:
I see. I misunderstood the code and missed the part that the subscriber is removed at the end and not going into infinite loop with the same subscriber...